### ISSUE TITLE
wrappers do not support event-driven out-of-order verdicts

### ISSUE THREAD

**Issue by binnacle** [chifflier/nfqueue-bindings#2]
Just finished writing an event-driven (i.e. IO::Select) 'nfqueue' application in 'perl' and discovered that $payload->set_verdict() produces a memory violation when called with a payload from an earlier, defunct $nfqueue->set_callback() callback.  In addition it was observed that any packet where set_verdict() is not called before the callback handler returns is automatically given a NF_ACCEPT verdict.  A guess is that the packet buffer is maintained on the call-stack and automatically destroyed when the handler exits.  Message payload information should be maintained on the heap rather than the stack.

According to https://home.regit.org/netfilter-en/using-nfqueue-and-libnetfilter_queue/ out-of-order 'nfqueue' verdicts are a usable design feature and should work fine.

For now the application is running with a hack where the 'nfqueue'  callback handler runs off an independent 'perl' select and the $nfqueue->process_pending() callback runs the master IO::Select->select() until each transaction completes.  Ugly as this largely defeats the asynchronous workflow implementation.

I'll take a look at this sooner of later, but I am not familiar with SWIG.  If a trivial adjustment to the wrapper specification will fix it and can be made by someone who knows how, that would be most helpful.

x86_64.  Built all libraries from latest GIT snapshots.  From-source build of 'perl. 5.24.0.  Tested and working (in one-at-a-time mode) on CentOS 6 2.6.32-642.1.1.el6.x86_64 kernel.

**Comment by binnacle** [comment]
Looking at `perl/libnetfilter_queue_perl.i` lines 73 and 79 appear the cause of the observed behavior:

```
   free(p);
.
.
.
   return nfq_set_verdict(qh, id, NF_ACCEPT, 0, NULL);
```

will replacing 79 with

```
   return 0;
```

and moving line 73 into the `set_verdict*()` methods found in `libnetfilter_queue.i` produce the desired result?

**Comment by binnacle** [comment]
This patch seems to fix the problem.  Also the second patch was an earlier necessity.

[nfqueue-bindings_async-patch.txt](https://github.com/chifflier/nfqueue-bindings/files/344680/nfqueue-bindings_async-patch.txt)
[nfqueue-bindings_build-patch.txt](https://github.com/chifflier/nfqueue-bindings/files/344681/nfqueue-bindings_build-patch.txt)

**Comment by binnacle** [comment]
Looking at the before/after generated code, it seems likely the addition of

```
%extend payload { 
   ~payload() {
      free($self);
   }
   ...
}
```

is superfluous.  Before:

```
XS(_wrap_delete_payload) {
   ...
   arg1 = (struct payload *)(argp1);
   free((char *) arg1);
   ...
}
```

and after:

```
SWIGINTERN void delete_payload(struct payload *self){
   free(self);
}
XS(_wrap_delete_payload) {
   ...
   arg1 = (struct payload *)(argp1);
   delete_payload(arg1);
   ...
}
```

are the same, no?  Were the `free(p)` calls removed by the patch in the perl and python .i files added to work-around a now-corrected SWIG bug?

**Comment by binnacle** [comment]
Last comment:  The explicit queue.i SWIG sources allocate the 'payload' object with malloc() and apply either SWIG_NewPointerObj (python) or SWIG_MakePtr (perl) to the result.  Is this the proper way to construct a 'payload' object?  The most recent Python commit inserts the `free(p)` to "fix a memory leak" and so I wonder if a more correct approach exists for creating payload objects such that memory is not leaked when they are destroyed?  Or perhaps SWIG does not generate the free() in the Python destructor?  Didn't generate Python wrappers here so I cannot see what SWIG does.  Also I wonder if that "memory leak" is simply delayed garbage collection?

**Comment by binnacle** [comment]
Dug at it some more and perhaps figured out the cause of the memory leak the last commit attempts to fix.

Suspect the reason 'payload' objects were not freed is that `SWIG_OWNER` should be set in the `SWIG_MakePtr` and `SWIG_NewPointerObj` calls the same way it is in `_wrap_new_payload()`.  The flag tells the wrapper code that it owns the memory and that `_wrap_delete_payload()` should release it by executing the destructor logic--without the flag the memory is not returned to the heap.

Here's the patch revised to specify `SWIG_OWNER` and to omit the superfluous explicit 'payload' destructor.  Running with it now and it seems ok, though with the application here it would be months before a memory leak would be noticeable.

[nfqueue-bindings_async-patch.txt](https://github.com/chifflier/nfqueue-bindings/files/344767/nfqueue-bindings_async-patch.txt)

**Comment by binnacle** [comment]
To make the asynchronous verdict + memory leak fix to retain the previous behavior, then

```
%extend payload { 
   ~payload() {
      nfq_set_verdict($self->qh, $self->id, NF_ACCEPT, 0, NULL);
      free($self);
   }
   ...
}
```

should do that.

**Comment by binnacle** [comment]
Tested it with the destructor change next above and it works perfectly.  This a) proves that setting SWIG_OWNER when allocating 'payload' objects fixes the memory leak and b) results in unchanged api semantics while providing proper asynchronous verdict behavior.  Final version of patch attached.

[nfqueue-bindings_async2-patch.txt](https://github.com/chifflier/nfqueue-bindings/files/345197/nfqueue-bindings_async2-patch.txt)

I prefer to run it without the auto-ACCEPT change.  In this approach, bugs in the application where the verdict is not set will show up as leaked packets sitting on the netfilter queue as seen with

```
cat /proc/net/netfilter/nfnetlink_queue
```

**Comment by chifflier** [comment]
Hi,
Thank you for the detailed bug report and patches. I'll have a look, here are a few comments:
- the default verdict was a choice, made to avoid problems if the handler forgot to provide one. Maybe this is not the best one, because it creates an unintended behavior. I'll probably remove it
- I think SWIG_OWNER should indeed be used (I can't remember why it was commented)
- the `free($self);` fragments looks weird to me. Is this part only written to retain the default verdict, or is it required to avoid the memory leak ?

Thanks,
Pierre

**Comment by binnacle** [comment]
> the default verdict was a choice, made to avoid problems if the handler forgot to provide one. Maybe this is not the best one, because it creates an unintended behavior. I'll probably remove it

I've been running without it and have had no problems.  Tested not setting verdict and this results in leaked/orphaned nfqueue entries which is what one would expect.  Unlikely to be missed.

> I think SWIG_OWNER should indeed be used (I can't remember why it was commented)

this was the cause of the memory leak mentioned in the most recent commit

> the free($self); fragments looks weird to me. Is this part only written to retain the default verdict, or is it required to avoid the memory leak ?

The explicit destructor is only to retain the default verdict behavior, but if one replaces the default destructor the free() must be added.  If the explicit one is removed, then the SWIG generated destructor will free the object.  I looked at the generated code when working on this.  Is easy to track down the dtor and see for yourself.  Also the relevant code fragments are show above in comment starting with "Looking at the before/after generated code".

Have stress-tested the script with the final patch less the explicit destructor and managed 300 NFQ transactions-per-second without tuning it much.  No leaks.  I'm planning to write another patch to expose more of the API, in particular the copy-size parameter in set_mode(), to libnfnetlink.c:nfnl_rcvbufsiz() so the user-land socket size can be increased, and want to expose the payload protocol value (ETH_TYPE_IP, ETH_TYPE_IPv6 etc.).  Maybe nfq_get_timestamp() and nfq_get_packet_hw() as well.
